src/tracer.h

```diff
@@ -174,22 +183,50 @@ class TraceManager {
         std::chrono::duration_cast<std::chrono::nanoseconds>(
             std::chrono::steady_clock::now().time_since_epoch());
 
-    opentelemetry::nostd::shared_ptr<otel_trace_api::Span> trace_span_{nullptr};
-
     std::unique_ptr<otel_trace_sdk::SpanExporter> exporter_;
 
     std::unique_ptr<otel_trace_sdk::SpanProcessor> processor_;
 
     std::shared_ptr<otel_trace_api::TracerProvider> provider_;
 
-    // Initializes Opentelemetry exporter, processor and provider
+    // OTel context to store spans, created in the current trace
+    opentelemetry::context::Context otel_context_;
+
+    // Initializes Opentelemetry exporter, processor, provider and context
     void InitTracer(const TraceConfigMap& config_map);
 
     // Starts a span with the provided timestamp
-    void InitSpan(const opentelemetry::common::SystemTimestamp& timestamp_ns);
-
-    // Ends the started span and cleans up tracer
-    void EndSpan();
+    opentelemetry::nostd::shared_ptr<otel_trace_api::Span> InitSpan(
+        std::string name,
+        const opentelemetry::common::SystemTimestamp& timestamp_ns,
+        const uint64_t& raw_timestamp_ns, bool is_root_span,
+        std::string parent_span_key = "");
+
+    // Ends the provided span
+    void EndSpanNow(
+        opentelemetry::nostd::shared_ptr<otel_trace_api::Span> span);
+
+    // Ends the provided span at specified steady timestamp
+    void EndSpan(
+        opentelemetry::nostd::shared_ptr<otel_trace_api::Span> span,
+        const uint64_t& raw_timestamp_ns);
+
+    // Reports TRITONSERVER_InferenceTraceActivity as event to
+    // the currently active span. If activity is an instance of
+    // `TRITONSERVER_TRACE_REQUEST_START` or
+    // `TRITONSERVER_TRACE_COMPUTE_START`,
+    // it starts a new request or compute span. For the request span it
+    // adds some triton related attributes, and adds this span to
+    // `otel_context_`. Alternativelly, if activity is
+    // `TRITONSERVER_TRACE_REQUEST_END` or `TRITONSERVER_TRACE_COMPUTE_END`,
+    // it ends the coresponding span.
+    void ReportToOpenTelemetry(
+        TRITONSERVER_InferenceTrace* trace,
+        TRITONSERVER_InferenceTraceActivity activity, uint64_t timestamp_ns);
+
+    // Returns the span key, for which the activity belongs
```

Comment:
Using span key and span name interchangeably appears to be confusing. Let's finalize to use one of the options.

Reply:
Span key and span name are different. Span name is what you will see in for example jaeger gui, when analyzing traces. Span key is what is used to store span entities in the context. Span's name is most likely will be different from span's key. I'll scan through to make sure everything is clear.

Reply:
Thanks for the clarification.